File: ruby/food-chain/food_chain.rb
class FoodChain

  def self.song
    [*0..7].map do |i|
      verse(i) + "\n"
    end.join("\n")
  end

  def self.animals
    [
      {animal: "fly"},
      {animal: "spider", description: "It wriggled and jiggled and tickled inside her."},
      {animal: "bird", description: "How absurd to swallow a bird!"},
      {animal: "cat", description: "Imagine that, to swallow a cat!"},
      {animal: "dog", description: "What a hog, to swallow a dog!"},
      {animal: "goat", description: "Just opened her throat and swallowed a goat!"},
      {animal: "cow", description: "I don't know how she swallowed a cow!"},
      {animal: "horse"}
    ]
  end

  def self.verse(index)
    current = animals[index]
    return last_verse(current) if current[:animal] == "horse"

    opening = self.opening(current[:animal])
    closing = self.closing
    compose_verse(opening, closing, current, index)
  end

  def self.compose_verse(opening, closing, current, index)
    [
      opening,
      description(current),
      middle_lines(index),
      closing
    ].flatten.reject(&:empty?).join("\n")
  end

  def self.description(current)
    current[:animal] == "fly" ? "" : current[:description]
  end

  def self.middle_lines(index)
    lines = [*0..index].reverse.map do |index|
      self.swallowed_animal(animals[index][:animal], animals[index-1][:animal])
    end
    trailing(lines)
  end

  def self.trailing(lines)
    verse = lines.empty? ? "" : lines
    spider_description(verse)
  end

  def self.swallowed_animal(current, previous)
    verse = current == "fly" ? "" : "She swallowed the #{current} to catch the #{previous}."
  end

  def self.spider_description(middle_verse)
    middle_verse.map do |line|
      line[36..-2] == ("spider") ? line.chop + format_spider_description : line
    end
  end

  def self.format_spider_description
    animals[1][:description].gsub("It", " that")
  end

  def self.opening(animal)
    "I know an old lady who swallowed a #{animal}."
  end

  def self.closing
    "I don't know why she swallowed the fly. Perhaps she'll die."
  end

  def self.last_verse(current)
    "I know an old lady who swallowed a #{current[:animal]}.
She's dead, of course!"
  end
end

File: javascript/food-chain/food-chain.js
var VERSES = [
  {
    animal: "fly"
  },
  {
    animal: "spider",
    intro: "It wriggled and jiggled and tickled inside her.",
    chased: "that wriggled and jiggled and tickled inside her"
  },
  {
    animal: "bird",
    intro: "How absurd to swallow a bird!"
  },
  {
    animal: "cat",
    intro: "Imagine that, to swallow a cat!"
  },
  {
    animal: "dog",
    intro: "What a hog, to swallow a dog!"
  },
  {
    animal: "goat",
    intro: "Just opened her throat and swallowed a goat!"
  },
  {
    animal: "cow",
    intro: "I don't know how she swallowed a cow!"
  },
  {
    animal: "horse"
  }
];

module.exports = {
  verse: function(index) {
    index  = index - 1
    var current = VERSES[index]
    var previous = VERSES.slice(0, index + 1);

    var lines = [opening(current.animal)]

    if (index == VERSES.length - 1) {
      lines.push("She's dead, of course!");
    } else {
      isNotLastVerse(lines, current, previous);
    }

    return lines.join("\n") + "\n";
  },
  verses: function(start, end) {
    var final_verse = [];
    for (var i = start; i <= end; i++) {
      final_verse.push(this.verse(i));
    }
    return final_verse.join("\n") + "\n";
  }
};

function opening (current) {
  return "I know an old lady who swallowed a " + current + ".";
}

function closing () {
  return "I don't know why she swallowed the fly. Perhaps she'll die.";
}

function isNotLastVerse(lines, current, previous) {
  var middle = middle_lines(previous);

  if (current.intro) {
    lines.push(current.intro);
  }

  if (middle) {
    lines.push(middle);
  }

  lines.push(closing());
}

function middle_lines (previous) {
  var lines = [];

  for (var i = 1; i < previous.length; i++) {
    var firstAnimal = previous[i].animal;
    var second = previous[i - 1];
    var secondAnimal = second.animal;

    line(firstAnimal, secondAnimal, second, lines)
  }
  return lines.reverse().join("\n")
}

function line(firstAnimal, secondAnimal, second, lines) {

    if (second.chased) {
      secondAnimal += " " + second.chased;
    }

    lines.push("She swallowed the " + firstAnimal + " to catch the " + secondAnimal + ".");
}
